tests: fix flaky tests with a race detector

The commit makes tests flaky:

https://github.com/tarantool/go-tarantool/commit/6a24a645122caeb44068d8ca93f3556e88630144

We need to fix the implementation of the `send` function to make it green again.
```
2026-02-03T11:27:08.4074453Z === RUN   TestGracefulShutdownWithReconnect
2026-02-03T11:27:10.0198138Z 2026/02/03 11:27:10 tarantool: reconnect (0/10) to 127.0.0.1:3014 failed: failed to dial: dial tcp 127.0.0.1:3014: connect: connection refused
2026-02-03T11:27:10.0272695Z 2026/02/03 11:27:10 Tarantool "127.0.0.1:3014" was unexpectedly terminated: <nil>
2026-02-03T11:27:10.6320039Z ==================
2026-02-03T11:27:10.6320323Z WARNING: DATA RACE
2026-02-03T11:27:10.6320562Z Write at 0x00c0002e4128 by goroutine 514:
2026-02-03T11:27:10.6321070Z   github.com/tarantool/go-tarantool/v3.(*Future).SetError()
2026-02-03T11:27:10.6324811Z       /home/runner/work/go-tarantool/go-tarantool/future.go:91 +0x104
2026-02-03T11:27:10.6325881Z   github.com/tarantool/go-tarantool/v3.(*futureList).clear()
2026-02-03T11:27:10.6326825Z       /home/runner/work/go-tarantool/go-tarantool/connection.go:253 +0xdc
2026-02-03T11:27:10.6327918Z   github.com/tarantool/go-tarantool/v3.(*Connection).closeConnection()
2026-02-03T11:27:10.6328924Z       /home/runner/work/go-tarantool/go-tarantool/connection.go:628 +0xa6c
2026-02-03T11:27:10.6329785Z   github.com/tarantool/go-tarantool/v3.(*Connection).Close()
2026-02-03T11:27:10.6330824Z       /home/runner/work/go-tarantool/go-tarantool/connection.go:429 +0xbb
2026-02-03T11:27:10.6331500Z   github.com/tarantool/go-tarantool/v3/test_helpers.isReady.deferwrap1()
2026-02-03T11:27:10.6332124Z       /home/runner/work/go-tarantool/go-tarantool/test_helpers/main.go:198 +0x33
2026-02-03T11:27:10.6332521Z   runtime.deferreturn()
2026-02-03T11:27:10.6332958Z       /opt/hostedtoolcache/go/1.25.6/x64/src/runtime/panic.go:589 +0x5d
2026-02-03T11:27:10.6333455Z   github.com/tarantool/go-tarantool/v3/test_helpers.StartTarantool()
2026-02-03T11:27:10.6334067Z       /home/runner/work/go-tarantool/go-tarantool/test_helpers/main.go:438 +0x1027
2026-02-03T11:27:10.6334591Z   github.com/tarantool/go-tarantool/v3/test_helpers.RestartTarantool()
2026-02-03T11:27:10.6335204Z       /home/runner/work/go-tarantool/go-tarantool/test_helpers/main.go:307 +0xc4
2026-02-03T11:27:10.6335783Z   github.com/tarantool/go-tarantool/v3_test.TestGracefulShutdownWithReconnect()
2026-02-03T11:27:10.6336420Z       /home/runner/work/go-tarantool/go-tarantool/shutdown_test.go:200 +0x2db
2026-02-03T11:27:10.6336790Z   testing.tRunner()
2026-02-03T11:27:10.6337233Z       /opt/hostedtoolcache/go/1.25.6/x64/src/testing/testing.go:1934 +0x21c
2026-02-03T11:27:10.6337858Z   testing.(*T).Run.gowrap1()
2026-02-03T11:27:10.6338334Z       /opt/hostedtoolcache/go/1.25.6/x64/src/testing/testing.go:1997 +0x44
2026-02-03T11:27:10.6338609Z 
2026-02-03T11:27:10.6338741Z Previous read at 0x00c0002e4128 by goroutine 576:
2026-02-03T11:27:10.6339132Z   github.com/tarantool/go-tarantool/v3.(*Connection).send()
2026-02-03T11:27:10.6339688Z       /home/runner/work/go-tarantool/go-tarantool/connection.go:1040 +0x76
2026-02-03T11:27:10.6340131Z   github.com/tarantool/go-tarantool/v3.(*Connection).Do()
2026-02-03T11:27:10.6340653Z       /home/runner/work/go-tarantool/go-tarantool/connection.go:1243 +0x84
2026-02-03T11:27:10.6341184Z   github.com/tarantool/go-tarantool/v3.(*Connection).newWatcherImpl.func1()
2026-02-03T11:27:10.6341776Z       /home/runner/work/go-tarantool/go-tarantool/connection.go:1488 +0x429
2026-02-03T11:27:10.6342043Z 
2026-02-03T11:27:10.6342140Z Goroutine 514 (running) created at:
2026-02-03T11:27:10.6342393Z   testing.(*T).Run()
2026-02-03T11:27:10.6342813Z       /opt/hostedtoolcache/go/1.25.6/x64/src/testing/testing.go:1997 +0x9d2
2026-02-03T11:27:10.6343193Z   testing.runTests.func1()
2026-02-03T11:27:10.6343628Z       /opt/hostedtoolcache/go/1.25.6/x64/src/testing/testing.go:2477 +0x85
2026-02-03T11:27:10.6343988Z   testing.tRunner()
2026-02-03T11:27:10.6344397Z       /opt/hostedtoolcache/go/1.25.6/x64/src/testing/testing.go:1934 +0x21c
2026-02-03T11:27:10.6344745Z   testing.runTests()
2026-02-03T11:27:10.6345161Z       /opt/hostedtoolcache/go/1.25.6/x64/src/testing/testing.go:2475 +0x96c
2026-02-03T11:27:10.6345511Z   testing.(*M).Run()
2026-02-03T11:27:10.6345909Z       /opt/hostedtoolcache/go/1.25.6/x64/src/testing/testing.go:2337 +0xed4
2026-02-03T11:27:10.6346366Z   github.com/tarantool/go-tarantool/v3_test.runTestMain()
2026-02-03T11:27:10.6346924Z       /home/runner/work/go-tarantool/go-tarantool/tarantool_test.go:4025 +0x35e
2026-02-03T11:27:10.6347372Z   github.com/tarantool/go-tarantool/v3_test.TestMain()
2026-02-03T11:27:10.6348075Z       /home/runner/work/go-tarantool/go-tarantool/tarantool_test.go:4029 +0x26
2026-02-03T11:27:10.6348439Z   main.main()
2026-02-03T11:27:10.6348678Z       _testmain.go:419 +0x171
2026-02-03T11:27:10.6348824Z 
2026-02-03T11:27:10.6349197Z Goroutine 576 (running) created at:
2026-02-03T11:27:10.6349599Z   github.com/tarantool/go-tarantool/v3.(*Connection).newWatcherImpl()
2026-02-03T11:27:10.6350180Z       /home/runner/work/go-tarantool/go-tarantool/connection.go:1456 +0x21c
2026-02-03T11:27:10.6350645Z   github.com/tarantool/go-tarantool/v3.(*Connection).dial()
2026-02-03T11:27:10.6351166Z       /home/runner/work/go-tarantool/go-tarantool/connection.go:520 +0xa5d
2026-02-03T11:27:10.6351627Z   github.com/tarantool/go-tarantool/v3.(*Connection).connect()
2026-02-03T11:27:10.6352163Z       /home/runner/work/go-tarantool/go-tarantool/connection.go:585 +0x8d
2026-02-03T11:27:10.6352567Z   github.com/tarantool/go-tarantool/v3.Connect()
2026-02-03T11:27:10.6353058Z       /home/runner/work/go-tarantool/go-tarantool/connection.go:388 +0xae9
2026-02-03T11:27:10.6353509Z   github.com/tarantool/go-tarantool/v3/test_helpers.isReady()
2026-02-03T11:27:10.6354081Z       /home/runner/work/go-tarantool/go-tarantool/test_helpers/main.go:191 +0x134
2026-02-03T11:27:10.6354599Z   github.com/tarantool/go-tarantool/v3/test_helpers.StartTarantool()
2026-02-03T11:27:10.6355200Z       /home/runner/work/go-tarantool/go-tarantool/test_helpers/main.go:438 +0x1027
2026-02-03T11:27:10.6355722Z   github.com/tarantool/go-tarantool/v3/test_helpers.RestartTarantool()
2026-02-03T11:27:10.6356307Z       /home/runner/work/go-tarantool/go-tarantool/test_helpers/main.go:307 +0xc4
2026-02-03T11:27:10.6356858Z   github.com/tarantool/go-tarantool/v3_test.TestGracefulShutdownWithReconnect()
2026-02-03T11:27:10.6357478Z       /home/runner/work/go-tarantool/go-tarantool/shutdown_test.go:200 +0x2db
2026-02-03T11:27:10.6358050Z   testing.tRunner()
2026-02-03T11:27:10.6358470Z       /opt/hostedtoolcache/go/1.25.6/x64/src/testing/testing.go:1934 +0x21c
2026-02-03T11:27:10.6358845Z   testing.(*T).Run.gowrap1()
2026-02-03T11:27:10.6359280Z       /opt/hostedtoolcache/go/1.25.6/x64/src/testing/testing.go:1997 +0x44
2026-02-03T11:27:10.6359611Z ==================
2026-02-03T11:27:11.6347306Z 2026/02/03 11:27:11 tarantool: reconnect (0/10) to 127.0.0.1:3014 failed: failed to dial: dial tcp 127.0.0.1:3014: connect: connection refused
2026-02-03T11:27:11.6424730Z 2026/02/03 11:27:11 Tarantool "127.0.0.1:3014" was unexpectedly terminated: <nil>
2026-02-03T11:27:11.6431433Z     testing.go:1617: race detected during execution of test
2026-02-03T11:27:11.6432289Z --- FAIL: TestGracefulShutdownWithReconnect (3.24s)
```

--- a/connection.go
+++ b/connection.go
@@ -1037,7 +1037,7 @@ func (conn *Connection) send(req Request, streamId uint64) *Future {
 
 	fut := conn.newFuture(req)
 
-	if fut.finished {
+	if fut.isFinished() {
 		conn.decrementRequestCnt()
 		return fut
 	}
@@ -1061,7 +1061,7 @@ func (conn *Connection) putFuture(fut *Future, req Request, streamId uint64) {
 	shard := &conn.shard[shardn]
 	shard.bufmut.Lock()
 
-	if fut.finished {
+	if fut.isFinished() {
 		shard.bufmut.Unlock()
 		return
 	}

--- a/future.go
+++ b/future.go
@@ -42,6 +42,13 @@ func (fut *Future) finish() {
 	fut.cond.Broadcast()
 }
 
+func (fut *Future) isFinished() bool {
+	fut.mutex.Lock()
+	defer fut.mutex.Unlock()
+
+	return fut.finished
+}
+
 // NewFuture creates a new empty Future for a given Request.
 func NewFuture(req Request) (fut *Future) {
 	fut = &Future{}

